### PR TITLE
feat: add importId to TrainScorerStep

### PR DESCRIPTION
**Pull request by tpmsr**
Adds `importId` to `TrainScorerStep` to support import of .dialog files with SwitchCondition nodes.

SwitchConditions may have clauses like `productType == WINDOWS` and `productType == OFFICE`, where `productType` can be set as the result of an API call.

When we generate a TrainDialog from a dialog tree with this type of structure, we will create an enum entity called `productType` to store each potential returned value of the API call, in this case `WINDOWS` and `OFFICE`.

We would like to set the appropriate enum entity value as a required condition on the next action generated from the dialog tree.  However, _actions_ are created as a secondary pass after a `TrainDialog` is generated.  Eg, there is an interactive or automated pass where each `TrainDialog` step is either matched to an existing action, or a new action is created.

This PR adds the ability to assign a unique id to each `TrainScorerStep`, which can be used to associate the scorer step with required conditions during action generation.

The PR also fixes some linter warnings.

### CODE REVIEW

File: src/Action.ts [hunk 1]
```diff
@@ -70,11 +70,11 @@ export class ActionBase {
     this.isTerminal = action.isTerminal
     this.isEntryNode = action.isEntryNode
     this.repromptActionId = action.repromptActionId
-    this.requiredEntitiesFromPayload = action.requiredEntitiesFromPayload || []
-    this.requiredEntities = action.requiredEntities || []
-    this.negativeEntities = action.negativeEntities || []
-    this.requiredConditions = action.requiredConditions || []
-    this.negativeConditions = action.negativeConditions || []
+    this.requiredEntitiesFromPayload = action.requiredEntitiesFromPayload
```
Comment (LarsLiden) [inline src/Action.ts]: Why this change?  This could potentially leave these fields in "undefined" state.   Would that be valid?  I'm assuming they were set to empty arrays as that's not valid

File: src/TrainDialog.ts [hunk 2]
```diff
@@ -45,6 +46,8 @@ export interface TrainScorerStep {
   importText?: string
   // Used for UI rendering only
   uiScoreResponse?: ScoreResponse
+  // If set, this action should be created with the specified condition(s).
```
Comment (LarsLiden) [inline src/TrainDialog.ts]: Does this field exist only for .dialog import?  If so, it doesn't feel right to add it to the core object if it's never used again.  Might be better to create a temporary object (i.e. perhaps a Map between the scorer step and required conditions)

File: src/Action.ts (hunk 1 above)
Comment (tpmsr) [reply src/Action.ts]: truthy

File: src/TrainDialog.ts (hunk 2 above)
Comment (tpmsr) [reply src/TrainDialog.ts]: Yes, currently this is only for .dialog import.  This seems like something that would be more generally useful, though, since action generation happens in a separate step from any kind of import.  Thoughts ?

File: src/Action.ts (hunk 1 above)
Comment (LarsLiden) [reply src/Action.ts]: Hmm...  As the constructor doesn't take a partial, theoretically one shouldn't be able to pass in undefined values, but once it is transpiled to JS all bets are off.  @mattmazzola - what do you think?

File: src/TrainDialog.ts (hunk 2 above)
Comment (mattmazzola) [reply src/TrainDialog.ts]: Yea, I would agree is confusing to add it to this object, but it's been done for a bunch of others `importHashes` and `enumValues` etc so idk.

I don't have much context on reasoning for as you seemed to have acknowledged `requiredConditions` on actions and in the description but yet still decided this field was needed.  It seems these TrainScorerSteps would have to be scrubbed later so this train score step with required conditions is never in the same object as actions which also had required conditions and they wonder which take precedent and why there are duplicates.

I see do alternatives:
1. Create `ImportScorerStep` only for import and has fields for import. After import is doens convert everything to normal.
2. Save `requiredConditions` on a temporary Action object linked by `labelAction` ID to associate them with scorerStep then do whatever

File: src/Action.ts (hunk 1 above)
Comment (mattmazzola) [reply src/Action.ts]: If you're confident that the constructor is never given a object that is not of type `ActionBase` object then it's likely good. I would try to find out why the `[]` were added in first place. Maybe the commit history has it

File: src/TrainDialog.ts (hunk 2 above)
Comment (LarsLiden) [reply src/TrainDialog.ts]: One difference with importHashes it that it continues to be used for the lifetime of the model as can add additional transcripts in the future.  .dialog import is one time only

File: src/Action.ts (hunk 1 above)
Comment (tpmsr) [reply src/Action.ts]: Well, this change was just trying to chip away at the wall of warnings we get when we build this repo.  Per the linter message, it seems like the "|| []" here will _never_ have any effect.

I'm guessing the pre-allocation of empty arrays in the field definitions are also irrelevant since they're always overwritten in the ctor.

File: src/TrainDialog.ts (hunk 2 above)
Comment (tpmsr) [reply src/TrainDialog.ts]: I'm not wedded to the placement here, was just trying to keep the conditions close to the thing that uses them and will be passed to the action creator.  Would you guys prefer a map/object to be set in the TrainDialog state ?

Comment (LarsLiden) [reply src/TrainDialog.ts]: At a minimum I suggest the variable name should be changed to indicate that it's only used for import.  Also, I think the server will throw an error as this is a unknown field.

Comment (tpmsr) [reply src/TrainDialog.ts]: This works on the server; see screenshots from the actual import in https://github.com/microsoft/ConversationLearner-UI/pull/1347

I'm fine moving this to a map in TrainDialog state, but I think we'd still need to add an `id` to the `TrainScorerStep` if we went with that approach.  I don't think we can uniquely identify a scorer step in the tree just by hashing its content.

**Comment by LarsLiden** [comment]
Interesting. Might check with Shahin to see if he relaxed schema enforcement.

**Comment by tpmsr** [comment]
> Interesting. Might check with Shahin to see if he relaxed schema enforcement.

Actually, I'm not clear on how this would impact the server ?  

We use this value in the creation of an action (`ActionBase`) but AFAIU the `TrainScorerStep` shouldn't ever hit the backend.

**Comment by LarsLiden** [comment]
Ah, right.  You are never saving the TrainDialog.  Got it.

**Comment by mattmazzola** [comment]
> We use this value in the creation of an action (ActionBase) but AFAIU the TrainScorerStep shouldn't ever hit the backend.

Wanted to clear up confusion. We send train dialogs the server all the time during creation and editing.  These train dialogs contain train scorer steps so they do hit the backend.

The issue is ensuring they aren't allowed to be sent with the `requiredConditions` property.  Perhaps you can put a runtime check into the client code to check that it's not left set when it shouldn't be.

**Comment by tpmsr** [comment]
Ok -- per offline conversations yesterday, I'm replacing the `Condition`s with an id that can be set during scorer step generation.

* We will set up a map in the `TrainDialogs` widget's component state to map between this id and the required conditions
  * We can't key this map by the `TrainScorerStep` since the .dialog importer does a deep copy of `TrainScorerStep`s before doing anything else with them
* We probably don't want to reuse `labelAction` to store this id since we don't have any other information to tell whether the `labelAction` should be used for condition matching or its regular purpose

@msft-shahins - FYI on this new field, it shouldn't be set in anything that goes to the backend

**Comment by tpmsr** [comment]
Sorry for being unclear - I meant that the TrainDialogs from this process should not hit the backend
________________________________
